clippy_lints/src/inherent_impl.rs

```diff
@@ -66,24 +91,43 @@ impl<'tcx> LateLintPass<'tcx> for MultipleInherentImpl {
 
             for impl_id in impl_ids.iter().map(|id| id.expect_local()) {
                 let impl_ty = cx.tcx.type_of(impl_id).instantiate_identity();
-                match type_map.entry(impl_ty) {
+                let hir_id = cx.tcx.local_def_id_to_hir_id(impl_id);
+                let criterion = match self.scope {
+                    InherentImplLintScope::Module => Criterion::Module(cx.tcx.parent_module(hir_id)),
+                    InherentImplLintScope::File => {
+                        if let Node::Item(&Item {
+                            kind: ItemKind::Impl(_impl_id),
+                            span,
+                            ..
+                        }) = cx.tcx.hir_node(hir_id)
+                        {
+                            Criterion::File(cx.tcx.sess.source_map().lookup_source_file(span.lo()).name.clone())
+                        } else {
+                            // We know we are working on an Impl, so the the pattern matching can
+                            // not fail
+                            unreachable!()
+                        }
+                    },
+                    InherentImplLintScope::Crate => Criterion::Crate,
+                };
+                match type_map.entry((impl_ty, criterion.clone())) {
                     Entry::Vacant(e) => {
                         // Store the id for the first impl block of this type. The span is retrieved lazily.
-                        e.insert(IdOrSpan::Id(impl_id));
+                        e.insert((IdOrSpan::Id(impl_id), criterion.clone()));
```

Comment:
There seems to be no need to insert the criterion as part of the value (while it is useful as part of the key). Can't you just keep inserting the `IdOrSpan` as was done before? Also, this would let you remove the `.clone()` and write `match type_map.entry((impl_ty, criterion)) {` above.